tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,261 @@
+            log.info("{}: Send out command: {} ", id, command);
+            controlChannel.println(command);
+        }
+    }
+
+    public class ErrorUpdater implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    log.error("{}: (stderr of the process):{}", id, line);
+                    errMsg = line;
+                }
+            } catch (IOException ioe) {
+                log.info("{}: Stderr of the process is closed.", id);
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "ErrorUpdater", e, doneFuture);
+            }
+        }
+    }
+
+    /**
+     * StatusUpdater reads ProcessWorker's stdout line by line. If the line is a JSON object, StatusUpdater updates
+     * the status with the JSON object.
+     */
+    public class StatusUpdater implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    try {
+                        JsonNode resp = JsonUtil.JSON_SERDE.readTree(line);
+                        if (resp.has("status")) {
+                            log.info("{}: Update the status: {}", id, resp.get("status").toString());
+                            status.update(resp.get("status"));
+                        }
+                        if (resp.has("log")) {
+                            log.info("{}: (stdout of the process): {}", id, resp.get("log").toString());
+                        }
+                        if (resp.has("error")) {
+                            JsonNode errNode = resp.get("error");
+                            if (errNode.getNodeType() == JsonNodeType.STRING) {
+                                errMsg = errNode.asText();
+                                log.error("{}: (stdout of the process):{}", id, errMsg);
+                            }
+                        }
+                    } catch (IOException e) {
+                        // not a JSON string
+                    }
+                }
+            } catch (IOException ioe) {
+                log.info("{}: Stdout f the process is closed.", id);
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "StatusUpdater", e, doneFuture);
+            }
+        }
+    }
+
+
+    @Override
+    public void stop(Platform platform) throws Exception {
+        if (!running.compareAndSet(true, false)) {
+            throw new IllegalStateException("ExternalCommandWorker is not running.");
+        }
+        log.info("{}: Deactivating ExternalCommandWorker.", id);
+        if (process != null) {
```

Comment:
There is a race condition here.  `ExternalCommandWorker#process` is used by the `ProcessMonitor` thread, and also here in a different thread.

Similarly, `controlChannel` is accessed by multiple threads and could be be used before it is initialized.

I think `process` and `controlChannel` need to be initialized and accessed under a lock.

I would suggest something like:
1. take lock
2. check if process is null.  if so then EXIT
3. create ControlCommand
4. release lock
5. call executor#awaitTermination for 1 minute
6. if the executor did not finish all tasks, then take the lock again, invoke process#destroy, release lock, call awaitTermination again

Reply:
Don't think this is necessary.  process.destroy() can be called multiple times as long as the process is created. When issuing command, the `println()` method has a lock inside already.

Reply:
It seems like the race conditions for `process` won't turn into any errors. It maybe is better to be explicit about the race conditions and have locks as to ease future maintainability of this task (especially by other contributors)

Reply:
Refactored the code:
* Added ExternalProcess, moved process and channel initialization works to ExternalProcess's constructor, also grouped ExternalProcess related control methods to ExternalPorcess too.
* Added an explicit lock when issuing a start/stop command.
* Only `stop` calls the destroy() now, so no race condition.